### PR TITLE
Less naive splitting for PaginatedChoiceState

### PR DESCRIPTION
**Pull request by gsvr**

### CODE REVIEW

**Comment by gsvr** [comment]
@justinvdm @hodgestar Please review.  Note I merged in #219 as I was expecting tests to change, so wait until that's merged, maybe, or just review the last commit?

This is probably not the most eloquent solution, but I didn't want to change too much of the existing code. Also, I did this over the weekend as a learning exercise, so no worries if you don't like it :)

File: test/test_states/test_choices.js [hunk 1]
```diff
@@ -450,9 +450,69 @@ describe("states.choice", function() {
...
+                            .check.reply.char_limit(opts.characters_per_page)
+                            .run();
+                    });
+            });
+        });
+
         describe("when the options per page is not fixed", function() {
             it("should dynamically split the choices", function() {
-                opts.question = 'Hello.',
+                opts.question = 'Hello.';
```
Comment (gsvr) [inline test/test_states/test_choices.js]: I changed this because it looked strange, but it doesn't seem to make any difference.

**Comment by hodgestar** [comment]
@gsvr I've merged #219. Can you merge develop into this branch?

File: lib/states/choices.js [hunk 2]
```diff
@@ -403,27 +416,49 @@ var PaginatedChoiceState = ChoiceState.extend(function(self, name, opts) {
...
+                .join('')
+                .length;
+        } else {
+            return self.characters_per_page - [
+                self.format_choice(self.back, self.choices.length),
+                self.format_choice(self.more, self.choices.length),
+                self.question_text + '\n']
+                .join('')
+                .length;
+        }
```
Comment (justinvdm) [inline lib/states/choices.js]: Don't we need to handle the last page too (and only take into account the 'back' choice, but not the 'more' choice)?

Comment (gsvr) [inline lib/states/choices.js]: That's what I thought at first, too, but n is based off the metadata.page_start so it didn't really end up working like that.  It's more a case of this function caters for starting at the beginning, and the other functions handling paging backwards and forwards from there. We can chat through it, maybe there's an error in my reasoning.

**Comment by gsvr** [comment]
@justinvdm @hodgestar I've addressed your comments if you want to have another look.

File: lib/states/choices.js (hunk 2 above)
Comment (justinvdm) [inline lib/states/choices.js]: Ah I see, thanks :)

File: test/test_states/test_choices.js (hunk 1 above)
Comment (hodgestar) [inline test/test_states/test_choices.js]: This change is good -- the comma worked a little accidentally I think.

File: lib/states/choices.js (hunk 2 above)
Comment (hodgestar) [inline lib/states/choices.js]: I think perhaps this discussion needs to be explained in a comment too.

**Comment by hodgestar** [comment]
@gsvr I'm +1 on this landing after we have a comment explaining the discussion you and @justinvdm had (because I was wondering the same thing). Would like a +1 from @justinvdm too though.

**Comment by justinvdm** [comment]
I made really minor comment and style fixes, but otherwise looks great! I'm going to go ahead and land once Travis is happy.

**Comment by justinvdm** [comment]
Travis builds are taking a while to start, I suspect Travis is quite busy atm, will land in the morning once Travis happy.

**Comment by hodgestar** [comment]
:+1:

@justinvdm Tx for doing the style and comment fixes.

**Pull request merged**